### PR TITLE
Add HTTP endpoint to show overrides (#1324)

### LINKED ISSUE #1324: Idea: Add HTTP endpoint to show overrides.
Like /ring, /user_stats etc.

### PR DESCRIPTION
**Pull request by ethervoid**
**What this PR does**:

This PR introduces a new endpoint that would retrieve the override configuration, per tenant, that is loaded as part of the runtime configuration

**Which issue(s) this PR fixes**:
Fixes #1324

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/cortex/modules.go
Comment (pstibrany) [inline pkg/cortex/modules.go:174]: This should be in `initRuntimeConfig` instead, after `t.RuntimeConfig` is set.

File: CHANGELOG.md
Comment (pstibrany) [inline CHANGELOG.md:24]: Enhancements go before BUGFIX. Also please add PR number to the end.

**Review by pstibrany — COMMENTED** [review]
Thanks for your PR! It looks good, I've left some comments to check before we can merge.

File: pkg/api/api.go
Comment (pstibrany) [reply pkg/api/api.go:181]: (Note that when issue #1324 was created, there was no "runtime config", only overrides. Which is why it doesn't mention it.)

File: pkg/api/handlers.go
Comment (pstibrany) [reply pkg/api/handlers.go:136]: Please note that this has now been done in https://github.com/cortexproject/cortex/pull/3645 too. To avoid duplicate work, it may be good idea to rebase on top of recent master, once https://github.com/cortexproject/cortex/pull/3645 is merged.

Comment (ethervoid) [reply pkg/api/handlers.go:136]: Great, thank you for pointing me to the new change. I'll wait until that PR is merged to make the rebase

File: pkg/api/api.go
Comment (ethervoid) [reply pkg/api/api.go]: Done. Now overrides is in the path `/runtime_config/overrides`

File: pkg/cortex/modules.go
Comment (ethervoid) [reply pkg/cortex/modules.go]: Change done

File: CHANGELOG.md
Comment (ethervoid) [reply CHANGELOG.md]: Done

**Comment by ethervoid** [comment]
@pstibrany I've made the requested changes but I'm waiting for the mentioned PR to be merged to do the last one referring to the refactoring of the YAML response code

File: pkg/api/api.go
Comment (pstibrany) [reply pkg/api/api.go]: Sorry for misunderstanding... what I meant is to expose entire “runtime configuration” object under `/runtime_config` path. Overrides are part of “runtime configuration”, so they will be automatically included by doing that.

Comment (pstibrany) [reply pkg/api/api.go]: By **“runtime configuration” object** I mean value returned by calling `t.RuntimeConfig.GetConfig()` method.

Comment (ethervoid) [reply pkg/api/api.go]: Oh, I see. Sorry for the noise, I've made the requested changes. Now we're exposing the runtime configuration through the `/runtime_config` endpoint.

File: docs/api/_index.md
Comment (pstibrany) [inline docs/api/_index.md:119]: Please add a note saying that if Cortex is not configured with any `-runtime-config.file`, then this will not be available.

File: pkg/api/api.go
Comment (pstibrany) [inline pkg/api/api.go:10]: Nit: this should go between cortex imports, and not be standalone.

File: pkg/api/api.go
Comment (pstibrany) [inline pkg/api/api.go:181]: I'd mention it includes overrides too: "Current Runtime Config (incl. Overrides)"

File: pkg/cortex/modules.go
Comment (pstibrany) [inline pkg/cortex/modules.go:174]: nit: extra empty line added

**Review by pstibrany — COMMENTED** [review]
Thank you, this looks good. One more detail that may be worth handling: if `runtimeCfgManager.GetConfig()` in the handler returns nil (which can happen if you define `-runtime-config.file` but file doesn't exist, then the output of `/runtime_config` is just "null". Perhaps some message like "runtime config file doesn't exist" would be more appropriate?

File: CHANGELOG.md
Comment (pstibrany) [inline CHANGELOG.md:28]: I would also mention that it includes overrides here.

**Comment by pstibrany** [comment]
Can we mention new `/runtime_config` endpoint in `docs/configuration/arguments.md` file ("Runtime Configuration file" section) too?

File: pkg/api/handlers.go
Comment (pracucci) [reply pkg/api/handlers.go:136]: I just merged #3645, so you can rebase `master` and just use that helper.

**Review by pracucci — APPROVED** [review]
Pretty cool! LGTM, modulo Peter's feedback.

In a **separate PR**, would be fantastic if you could also add the `mode=diff` support like in https://github.com/cortexproject/cortex/pull/3645 because it's a bit hard to read the output due to the large quantity of default valued fields. If you don't, please open an issue to leave it for someone else 🙏

**Comment by pstibrany** [comment]
> In a **separate PR**, would be fantastic if you could also add the `mode=diff` support like in #3645 because it's a bit hard to read the output due to the large quantity of default valued fields. If you don't, please open an issue to leave it for someone else 🙏

This doesn’t apply to runtime config. `/runtime_config` only shows your changes (yaml file) nothing else.

**Comment by ethervoid** [comment]
I"ve been checking about the diff option and looks like we're setting the default values [here](https://github.com/cortexproject/cortex/blob/2dd7692140fc1777acef736993d89211ee826bd2/pkg/cortex/modules.go#L171) to have it in the YAML unmarshal operation so I think we can add a diff option but it would be only for the `overrides` part

**Comment by pstibrany** [comment]
> This doesn’t apply to runtime config. `/runtime_config` only shows your changes (yaml file) nothing else.

I am wrong about this. I tested this PR earlier today, but without actually setting any overrides 🤦 and just other parts of runtime config.

> I"ve been checking about the diff option and looks like we're setting the default values [here](https://github.com/cortexproject/cortex/blob/2dd7692140fc1777acef736993d89211ee826bd2/pkg/cortex/modules.go#L171) to have it in the YAML unmarshal operation so I think we can add a diff option but it would be only for the `overrides` part

I think diff makes sense for entire runtime config, but is mostly interesting for overrides. However it is also somewhat tricky for overrides... when diffing, each per-tenant override should be diffed against default limits individually, and not the whole structure.

As Marco suggested, let's leave that for another PR to get this merged.

File: pkg/api/handlers.go
Comment (pstibrany) [inline pkg/api/handlers.go:229]: This isn’t error, it just means that overrides file is empty or doesn’t exist. Both are perfectly fine. Writing text message with code 200 is preferred.

File: docs/api/_index.md
Comment (pstibrany) [inline docs/api/_index.md:133]: `-runtime-config.file` option must be set for this endpoint to be available, but actual file doesn’t need to exist or be valid. But that’s just nitpicking.

**Review by pstibrany — APPROVED** [review]
Thanks for patiently addressing our feedback. Let’s fix the nil case, and merge.

**Comment by ethervoid** [comment]
@pstibrany I think that's all. I want to thank you for your amazing patiente guiding me and providing really valuable feedback. I've learn a lot about the the project and the way the thins are done so I can be more accurate for future PRs :-)

**Comment by pstibrany** [comment]
Thank you for addressing my feedback!

**Pull request merged**
